### PR TITLE
Vp/min 32 borrow cap mode

### PR DESCRIPTION
**Pull request by dev-vadim**
**Description:**

Implemented borrow cap for each pool to disallow borrow if total borrowed amount will be greater that cap.

**Changes in storage:**

Added borrow_cap to ControllerData

**Test scenarios/Use cases:**

When borrow cap mode is enabled, borrow operation is disallowed if total_borrowed will be greater than borrow cap

### CODE REVIEW

File: pallets/controller/src/lib.rs [hunk 1]
```diff
@@ -379,6 +401,72 @@ pub mod module {
 			Self::deposit_event(Event::CollateralFactorChanged);
 			Ok(().into())
 		}
+
+		/// Enable borrow cap mode.
+		///
+		/// The dispatch origin of this call must be Administrator.
+		/// Borrow cap value must be in range 0..1_000_000_000_000_000_000_000_000
```
Comment (lut1k) [inline pallets/controller/src/lib.rs:408]: Hmm ..
If I need to set borrow_cap = 100 USD?
In Current implementation 100USD = 100 * 10 ^ 18

File: pallets/controller/src/mock.rs [hunk 2]
```diff
@@ -149,12 +149,18 @@ impl minterest_model::Config for Runtime {
...
+}
+
 impl Config for Runtime {
 	type Event = Event;
 	type LiquidityPoolsManager = liquidity_pools::Module<Runtime>;
+	type MaxBorrowCap = MaxBorrowCap;
 }
 
 pub const MAX_MEMBERS: u8 = 16;
+pub const MAX_BORROW_CAP: Balance = 1_000_000_000_000_000_000_000_000;
```
Comment (lut1k) [inline pallets/controller/src/mock.rs:163]: This line suggests that MAX_BORROW_CAP = 1 USD.
It is not true

File: pallets/controller/src/lib.rs [hunk 3]
```diff
@@ -44,6 +44,9 @@ pub struct ControllerData<BlockNumber> {
```
Comment (Pavel-Blaize) [inline pallets/controller/src/lib.rs:44]: Please, upgrade pallet version to 0.6.2 since storage changes are made

**Review by Pavel-Blaize — APPROVED** [review]

File: pallets/controller/src/lib.rs [hunk 4]
```diff
@@ -133,6 +146,8 @@ pub mod module {
 		DepositedInsurance(CurrencyId, Balance),
 		/// Insurance balance redeemed: \[pool_id, amount\]
 		RedeemedInsurance(CurrencyId, Balance),
+		/// Borrow cap changed: \[pool_id, new_cap\]
+		BorrowCapChanged(CurrencyId, Balance),
```
Comment (lut1k) [inline pallets/controller/src/lib.rs:150]: If you use the Option type, then change:
```BorrowCapChanged(CurrencyId, Option<Balance>)```

File: pallets/controller/src/lib.rs [hunk 5]
```diff
@@ -378,6 +394,34 @@ pub mod module {
...
+				T::LiquidityPoolsManager::pool_exists(&pool_id),
+				Error::<T>::PoolNotFound
+			);
+
+			ensure!(
+				borrow_cap >= Balance::zero() && borrow_cap <= T::MaxBorrowCap::get(),
+				Error::<T>::InvalidBorrowCap
+			);
+			ControllerDates::<T>::mutate(pool_id, |r| r.borrow_cap = borrow_cap);
+			Self::deposit_event(Event::BorrowCapChanged(pool_id, borrow_cap));
```
Comment (lut1k) [inline pallets/controller/src/lib.rs:421]: If you use the Option type, then change:
```
ControllerDates::<T>::try_mutate(pool_id, |data| -> DispatchResult {
	if let Some(cap) = borrow_cap {
		ensure!(
			cap >= Balance::zero() && cap <= T::MaxBorrowCap::get(),
			Error::<T>::InvalidBorrowCap
		);
	}
	data.borrow_cap = borrow_cap;
	Ok(())
});

Self::deposit_event(Event::BorrowCapChanged(pool_id, borrow_cap));
```

File: pallets/controller/src/lib.rs [hunk 6]
```diff
@@ -614,7 +658,21 @@ impl<T: Config> Pallet<T> {
...
+			let new_total_borrows = pool_total_borrowed
+				.checked_add(borrow_amount)
+				.ok_or(Error::<T>::NumOverflow)?;
+			let new_total_borrows_in_usd = Rate::from_inner(new_total_borrows)
+				.checked_mul(&oracle_price)
+				.map(|x| x.into_inner())
+				.ok_or(Error::<T>::NumOverflow)?;
+
+			ensure!(new_total_borrows_in_usd <= borrow_cap, Error::<T>::BorrowCapReached);
+		}
```
Comment (lut1k) [inline pallets/controller/src/lib.rs:675]: If you use the Option type, you can write in the Rust style:
...
```
if let Some(borrow_cap) = Self::get_borrow_cap(underlying_asset_id) {
	ensure!(new_total_borrows_in_usd <= borrow_cap, Error::<T>::BorrowCapReached);
}
```

File: pallets/controller/src/lib.rs [hunk 7]
```diff
@@ -1013,6 +1071,11 @@ impl<T: Config> Pallet<T> {
 	fn get_insurance_factor(pool_id: CurrencyId) -> Rate {
 		Self::controller_dates(pool_id).insurance_factor
 	}
+
+	/// Gets the borrow cap amount
+	fn get_borrow_cap(pool_id: CurrencyId) -> Balance {
```
Comment (lut1k) [inline pallets/controller/src/lib.rs:1076]: If you use the Option type:
``` fn get_borrow_cap(pool_id: CurrencyId) -> Option<Balance> {} ```

**Review by lut1k — APPROVED** [review]

**Review by Apollo50 — APPROVED** [review]

**Pull request merged**